### PR TITLE
[incidentist] Extract report generation logic into separate package

### PR DESCRIPTION
**Pull request by kkhor-datadog**
Extracts the core report generation logic into a separate package, so that it can eventually be reused in other repositories. There are no functional changes expected.

Auxiliary changes include:
- Fix use of deprecated Pagerduty client APIs e.g. replacing `client.ListIncidents` with `client.ListIncidentsWithContext`
- Explicitly parse the environment variables `DD_API_KEY` and `DD_APP_KEY` for the Datadog API and app keys and use these to explicitly construct a Datadog API client context, instead of letting the Datadog API client [handle the environment variable parsing on its own](https://github.com/DataDog/datadog-api-client-go/blob/cdaf8699f0f7f653756f6af05bb62068cccf0a41/api/datadog/configuration.go#L548). We want to do this so that we can explicitly specify the Datadog URL which the API client will query.

Tested by generating a report from the existing code and the new code, and verified that there were no diffs, as expected.

### CODE REVIEW

File: report/generate_report.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	Until      string
+	TagFilters []string
+	AuthToken  string
+	Urgency    string
+	Replace    []string
+	DdApiKey   string
+	DdAppKey   string
+}
+
+func GenerateReport(request GenerateReportRequest) (string, error) {
```
Comment (kkhor-datadog) [inline report/generate_report.go]: This is the core public API that will be exposed by the package

File: main.go [hunk 2]
```diff
@@ -50,410 +41,34 @@ func main() {
...
-	return regexReplace
-}
-
-func newIntPtr(v int) *int {
-	return &v
-}
-
-func newUInt64(v uint64) *uint64 {
-	return &v
-}
```
Comment (kkhor-datadog) [inline main.go:184]: These pointer helpers are unused

File: main.go [hunk 3]
```diff
@@ -50,410 +41,34 @@ func main() {
...
-					md.unordered(3, fmt.Sprintf("**%s**: %s", n.userEmail, n.content))
-				} else {
-					md.unordered(3, n.content)
-				}
-			}
-			md.br()
-		}
-		md.unordered(2, "**Action taken**: "+filloutPlaceholder)
-		md.unordered(2, "**Follow-up**: "+filloutPlaceholder)
+	request := report.GenerateReportRequest{
```
Comment (xornivore) [inline main.go]: Nit: what do you think about avoiding `Report` in the func name since the package is named `report`?

File: main.go [hunk 4]
```diff
@@ -50,410 +41,34 @@ func main() {
...
-	incResp, err := client.ListIncidents(pagerduty.ListIncidentsOptions{
-		Limit:     1000,
-		TeamIDs:   []string{teamID},
-		Since:     since,
-		Until:     until,
-		Urgencies: []string{*urgency},
-		// Includes:  []string{"users", "acknowledgers", "acknowledgers"},
-	})
 
+	report, err := report.GenerateReport(request)
```
Comment (xornivore) [inline main.go]: Maybe it could be just `Generate` - see the comment about report/Report.

**Review by xornivore — COMMENTED** [review]
Thank you for doing this work - had it as a TODO for myself for a while, so happy to see it done! @kkhor-datadog 🔥

**Review by xornivore — APPROVED** [review]
🔥 🙇

**Pull request merged**